### PR TITLE
🔧 Create file storage for bolt

### PR DESCRIPTION
**Pull request by barlock**
**Related PRs**  
This PR is not dependent on any other PR

**What does this PR do?**  
Creates file storage for bolt. Is quite useful for restarting with nodemon.

**Description of Changes**  
Modified the default [MemoryStore](https://github.com/slackapi/bolt/blob/master/src/conversation-store.ts#L20) and took tests from it to ensure compatibility.

**What gif most accurately describes how I feel towards this PR?**

### CODE REVIEW

File: packages/bolt-storage-file/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Bolt File Storage
```
Comment (tteltrab) [inline packages/bolt-storage-file/README.md]: don't love the name "storage-file" for this package. no immediate suggestions - maybe something combination of `dev`, `store`, `persist`, `file`, `state`. like `bolt-dev-persist` or `bolt-state-store`. `bolt-file-store` sounds like you're storing files almost, though, which doesn't make a ton of sense.

File: packages/bolt-storage-file/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+Bolt [conversation store](https://slack.dev/bolt/concepts#conversation-store) that uses your file system as a database. Useful for retaining conversation state through restarts of your App in development.
```
Comment (tteltrab) [inline packages/bolt-storage-file/README.md]: let's write this as a sentence and maybe give a little more detail on the why/problem being solved? no strong feelings, here's a stab

```suggestion
Tool to maintain a Bolt [conversation store](https://slack.dev/bolt/concepts#conversation-store) in development using your file system as a database. Improves developer experience by persisting conversation state through App restarts.
```

File: packages/bolt-storage-file/README.md (hunk 1 above)
Comment (tteltrab) [reply packages/bolt-storage-file/README.md]: also, probably worth providing a simple helper to clean the persistent storage? maybe as a separate PR.

File: packages/bolt-storage-file/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+```
+
+## Usage
+
+```typescript
+import { App } from '@slack/bolt';
+import FileStore from '@slack-wrench/bolt-storage-file';
+
+const databasePath = '.db'; // defaults to ".db"
+const fileStore = new FileStore(databasePath);
```
Comment (tteltrab) [inline packages/bolt-storage-file/README.md]: wouldn't the happiest path jusrt be 
```js
const fileStore = new FileStore(); 
```

? 

my guess is most people wouldn't customize. don't feel strongly here, just wondering how easy we want the initial example to be.

Could have a different header showing how to customize the usage by passing in a `databasePath`

File: packages/bolt-storage-file/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    if (entry.expiresAt !== undefined && Date.now() > entry.expiresAt) {
+      // release the memory
+      await this.db.del(conversationId);
+
+      throw new Error('Conversation expired');
+    }
+
+    return entry.value;
+  }
```
Comment (tteltrab) [inline packages/bolt-storage-file/src/index.ts]: shouldn't we also error if the entry isn't found? bolt does this. 

https://github.com/slackapi/bolt/blob/master/src/conversation-store.ts#L39

File: packages/bolt-storage-file/src/index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+describe('FileStore', () => {
+  const defaultDbPath = path.join(process.cwd(), `.db`);
+  let dbPath: string;
+  let store: FileStore;
+  let id: string;
+  let initialState: Record<string, any>;
+
+  beforeEach(() => {
+    // level-db doesn't work when being quickly recreated in the same dir
```
[the remark this answers is not in the document]
Comment (barlock) [reply packages/bolt-storage-file/src/index.spec.ts]: Not an obvious issue I could find, I'm also not sure where the issue is. Like which if the level packages is causing the issue. The whole `(level|rock)(up|down|sideways|backwards)` ecosystem is just, a lot.

I looked around for another package, but nothing seemed to do it as simply and be as featureful. Sense it's mainly intended to be a used for dev, I figured it didn't really matter so much in practice.

File: packages/bolt-storage-file/src/index.spec.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    dbPath = path.join(__dirname, `.db-${shortid.generate()}`);
+    store = new FileStore(dbPath);
+
+    id = 'CONVERSATION_ID';
```
[the remark this answers is not in the document]
Comment (barlock) [reply packages/bolt-storage-file/src/index.spec.ts]: It's not, good spot.

File: packages/bolt-storage-file/README.md (hunk 1 above)
Comment (barlock) [reply packages/bolt-storage-file/README.md]: > don't love the name "storage-file"

Same. My intent was to play prefix games for sorting and the like.

For example, in production we will likely need to write a "bolt-storage-postgres". I'm both surprised and not at all that I couldn't find any other existing art here. 

I could work `convo` store in there too, but like. That seems weird, I'm also trying to use it for interactions which isn't a convo technically?

Comment (barlock) [reply packages/bolt-storage-file/README.md]: > probably worth providing a simple helper to clean the persistent storage? maybe as a separate PR.

Seems interesting, I'm not sure I see what dx your looking to help.

My current process is just `rm -rf .db`

File: packages/bolt-storage-file/README.md (hunk 2 above)
Comment (barlock) [reply packages/bolt-storage-file/README.md]: > Tool to maintain a Bolt

😛 Na it's litterally a conversation store.

https://github.com/IBM/slack-wrench/pull/2/files#diff-7dcefd9507c82601c7c829c82b1a3005R10

I'll take a stab at it though.

File: packages/bolt-storage-file/README.md (hunk 3 above)
Comment (barlock) [reply packages/bolt-storage-file/README.md]: Having a whole section for `path` seems unnecessary to me. I'll try something else to communicate it.

File: packages/bolt-storage-file/src/index.ts (hunk 4 above)
Comment (barlock) [reply packages/bolt-storage-file/src/index.ts]: We should. Here's the test case https://github.com/IBM/slack-wrench/pull/2/files#diff-4b729154570b2a7cb86c2298a52b5eb3R53

They had to code it in by hand b/c memory store. level does it for us.

File: packages/bolt-storage-file/README.md (hunk 1 above)
Comment (tteltrab) [reply packages/bolt-storage-file/README.md]: kind of a small thing, but that currently requires people both knowing how this is implemented and where the file is/its name.

**Pull request merged**